### PR TITLE
executor: track the mem usage of IndexMergeReader

### LINKED ISSUE #30208: Didn't tracker memory usage of IndexMergeReader
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

    drop table if exists t1;
    create table t1(c1 varchar(100), c2 varchar(100), c3 varchar(100), key(c1), key(c2));
    load data local infile '/home/guo-shaoge/str.sql' into table t1 fields terminated by ' ';
    select /*+ use_index_merge(t1) */ * from t1 where c1 < '10' or c2 < '10'; 

### 2. What did you expect to see? (Required)
The memory column is 312Bytes.

    mysql> explain analyze select /*+ use_index_merge(t1) */ * from t1 where c1 < '10' or c2 < '10';
    +-------------------------------+---------+---------+-----------+------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------------------------+-----------+------+
    | id                            | estRows | actRows | task      | access object          | execution info                                                                                                                                                         | operator info
                       | memory    | disk |
    +-------------------------------+---------+---------+-----------+------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------------------------+-----------+------+
    | IndexMerge_8                  | 5542.21 | 0       | root      |                        | time:3.98ms, loops:1,                                                                                                                                                  |
                       | 312 Bytes | N/A  |
    | ├─IndexRangeScan_5(Build)     | 3323.33 | 0       | cop[tikv] | table:t1, index:c1(c1) | time:0s, loops:0, cop_task: {num: 1, max: 3.67ms, proc_keys: 0, tot_proc: 3ms, rpc_num: 1, rpc_time: 3.66ms, copr_cache_hit_ratio: 0.00}, tikv_task:{time:0s, loops:0} | range:[-inf,"10"), keep order:false, stats:pseudo | N/A       | N/A  |
    | ├─IndexRangeScan_6(Build)     | 3323.33 | 0       | cop[tikv] | table:t1, index:c2(c2) | time:0s, loops:0, cop_task: {num: 1, max: 3.65ms, proc_keys: 0, tot_proc: 3ms, rpc_num: 1, rpc_time: 3.63ms, copr_cache_hit_ratio: 0.00}, tikv_task:{time:0s, loops:0} | range:[-inf,"10"), keep order:false, stats:pseudo | N/A       | N/A  |
    | └─TableRowIDScan_7(Probe)     | 5542.21 | 0       | cop[tikv] | table:t1               |                                                                                                                                                                        | keep order:false, stats:pseudo                    | N/A       | N/A  |
    +-------------------------------+---------+---------+-----------+------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------------------------+-----------+------+
### 3. What did you see instead (Required)
The memory column is N/A

    mysql> explain analyze select /*+ use_index_merge(t1) */ * from t1 where c1 < '10' or c2 < '10';
    +-------------------------------+---------+---------+-----------+------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------------------------+--------+------+
    | id                            | estRows | actRows | task      | access object          | execution info                                                                                                                                                         | operator info                                     | memory | disk |
    +-------------------------------+---------+---------+-----------+------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------------------------+--------+------+
    | IndexMerge_8                  | 5542.21 | 0       | root      |                        | time:4.3ms, loops:1,                                                                                                                                                   |                                                   | N/A    | N/A  |
    | ├─IndexRangeScan_5(Build)     | 3323.33 | 0       | cop[tikv] | table:t1, index:c1(c1) | time:0s, loops:0, cop_task: {num: 1, max: 4.03ms, proc_keys: 0, tot_proc: 3ms, rpc_num: 1, rpc_time: 4ms, copr_cache_hit_ratio: 0.00}, tikv_task:{time:0s, loops:0}    | range:[-inf,"10"), keep order:false, stats:pseudo | N/A    | N/A  |
    | ├─IndexRangeScan_6(Build)     | 3323.33 | 0       | cop[tikv] | table:t1, index:c2(c2) | time:0s, loops:0, cop_task: {num: 1, max: 3.93ms, proc_keys: 0, tot_proc: 3ms, rpc_num: 1, rpc_time: 3.91ms, copr_cache_hit_ratio: 0.00}, tikv_task:{time:0s, loops:0} | range:[-inf,"10"), keep order:false, stats:pseudo | N/A    | N/A  |
    | └─TableRowIDScan_7(Probe)     | 5542.21 | 0       | cop[tikv] | table:t1               |                                                                                                                                                                        | keep order:false, stats:pseudo                    | N/A    | N/A  |
    +-------------------------------+---------+---------+-----------+------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------------------------+--------+------+
### 4. What is your TiDB version? (Required)

    mysql> select tidb_version();
    +----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
    | tidb_version()
                              |
    +----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
    | Release Version: v5.4.0-alpha-241-g7a8c7f89ce
    Edition: Community
    Git Commit Hash: 7a8c7f89cebe0e5596415a6106b1345cec9cfa27
    Git Branch: master
    UTC Build Time: 2021-11-28 02:32:15
    GoVersion: go1.16.4
    Race Enabled: false
    TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
    Check Table Before Drop: false |
    +----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
    1 row in set (0.00 sec)

### PR DESCRIPTION
**Pull request by guo-shaoge**
Signed-off-by: guo-shaoge <shaoge1994@163.com>

### What problem does this PR solve?

Issue Number: close #30208

Problem Summary: In IndexMergeReaderExecutor, the memTracker is nil so we cannot track it.

### What is changed and how it works?

Init IndexMergeReaderExecutor.memTracker, it will track the follow parts of memory usage:
1. Track the mem usage of TableScanWorker. It includes: the chunk returned from TableReaderExec and handles returned from PartialIndexWorker/PartialTableWorker. And it will be released when the chunk is returned to caller.
2. Track the mem usage of kv requestBuilder(which is used in `PartialIndexWorker`). It includes: the rpc response size of index scan and will be released when the `copIteratorWorker` received the result.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
executor: track the mem usage of IndexMergeReader
```

### CODE REVIEW

**Comment by fuzhe1989** [comment]
LGTM (though effectless)

File: executor/index_merge_reader_test.go [hunk 1]
```diff
@@ -170,3 +172,38 @@ func (s *testSuite1) TestPartitionTableRandomIndexMerge(c *C) {
...
+	tk.MustQuery("select /*+ use_index_merge(t1) */ * from t1 where c1 > 1 or c2 > 1")
+
+	newMaxUsage := memTracker.MaxConsumed()
+	c.Assert(newMaxUsage, Greater, oriMaxUsage)
+
+	res := tk.MustQuery("explain analyze select /*+ use_index_merge(t1) */ * from t1 where c1 > 1 or c2 > 1")
+	c.Assert(len(res.Rows()), Equals, 4)
+	// Parse "xxx KB" and check it's greater than 0.
+	memStr := res.Rows()[0][7].(string)
+	re, err := regexp.Compile("[0-9]+ KB")
```
Comment (XuHuaiyu) [inline executor/index_merge_reader_test.go:203]: Will `0 KB` also matches this regexp?

**Review by XuHuaiyu — COMMENTED** [review]
``` sql
func (w *indexMergeTableScanWorker) executeTask(ctx context.Context, task *lookupTableTask) error {

	defer func (){
		w.memTracker.Consume(-w.memTracker.BytesConsumed())
	}()
	...
```

File: executor/index_merge_reader_test.go (hunk 1 above)
Comment (guo-shaoge) [reply executor/index_merge_reader_test.go:203]: Yeah. But line 208 will check the mem usage must not be zero.

**Review by XuHuaiyu — APPROVED** [review]
LGTM

**Review by wshwsh12 — APPROVED** [review]

**Pull request merged**